File: ai-service/services/teaching_units.py
import os
from typing import Dict, Any, Optional, List
from openai import OpenAI
import asyncio
import json

_client = None

def get_openai_client():
    global _client
    if _client is None:
        api_key = os.getenv("OPENAI_API_KEY")
        if not api_key:
            return None
        _client = OpenAI(api_key=api_key)
    return _client


async def generate_teaching_unit(
    subject: str,
    topic: str,
    subtopics: Optional[List[str]] = None,
    weak_areas: Optional[List[str]] = None,
    pyq_data: Optional[List[Dict]] = None
) -> Dict[str, Any]:
    client = get_openai_client()

    if not client:
        return get_fallback_teaching_unit(subject, topic)

    try:
        context_parts = [
            f"Subject: {subject}",
            f"Topic: {topic}"
        ]
        if subtopics:
            context_parts.append(f"Key subtopics: {', '.join(subtopics)}")
        if weak_areas:
            context_parts.append(f"Student weak areas: {', '.join(weak_areas)}")
        if pyq_data:
            pyq_summary = []
            for q in pyq_data[:10]:
                pyq_summary.append(f"- {q.get('subtopic', 'General')}: {q.get('stem', '')[:100]}")
            context_parts.append(f"Sample PYQs from this topic:\n" + "\n".join(pyq_summary))

        context = "\n".join(context_parts)

        prompt = f"""You are a NEET-PG exam preparation expert and medical educator. Generate a comprehensive Teaching Unit for the following topic.

{context}

Generate the following in JSON format:
{{
  "concept_core_block": [
    // 8-12 core concept entries. Each entry is an object with:
    // "title": short concept name
    // "type": one of "definition", "mechanism", "differentiation", "classification", "principle"
    // "content": clear, concise explanation (2-4 sentences)
    // "high_yield": boolean indicating if this is frequently tested in NEET-PG
    // Focus on definitions, mechanisms, differentiations, and classifications.
  ],
  "comparison_tables": [
    // 2-4 high-yield comparison tables. Each table is an object with:
    // "title": what is being compared (e.g., "Type 1 vs Type 2 RTA")
    // "columns": array of column headers (entity names being compared)
    // "rows": array of objects with "feature" and "values" (array matching columns)
    // Include 5-8 comparison points per table.
    // Choose the most commonly confused or compared entities.
  ],
  "clinical_scenarios": [
    // 3-5 clinical application scenarios. Each is an object with:
    // "scenario": a brief clinical vignette (2-3 sentences)
    // "key_concept": the concept being tested
    // "expected_answer": what the student should recognize/conclude
    // "teaching_point": one-line takeaway
    // These should mimic NEET-PG clinical question style.
  ],
  "numerical_recall_points": [
    // 5-8 numerical/value-based facts to memorize. Each is an object with:
    // "fact": the numerical fact (e.g., "Normal serum sodium: 135-145 mEq/L")
    // "context": why this value matters clinically
    // "mnemonic": optional memory aid (null if none)
    // Include lab values, dosages, cutoffs, staging criteria.
  ],
  "trap_patterns": [
    // 4-6 common exam trap patterns. Each is an object with:
    // "trap": description of the trap (e.g., "Confusing X with Y because...")
    // "why_tempting": why students fall for it
    // "correct_approach": how to avoid the trap
    // "related_subtopic": the subtopic this trap relates to
    // Based on common PYQ distractor patterns.
  ]
}}

Rules:
- All content must be medically accurate and NEET-PG relevant.
- Prioritize high-yield, frequently tested concepts.
- Comparison tables should cover the most commonly confused pairs.
- Clinical scenarios should mimic real NEET-PG question style.
- Trap patterns should reflect actual PYQ distractor strategies.
- Return ONLY valid JSON, no markdown or extra text."""

        loop = asyncio.get_event_loop()
        response = await loop.run_in_executor(None, lambda: client.chat.completions.create(
            model=os.getenv("AI_MODEL", "gpt-4o-mini"),
            messages=[
                {"role": "system", "content": "You are a NEET-PG medical education expert. Always return valid JSON only."},
                {"role": "user", "content": prompt}
            ],
            temperature=0.7,
            max_tokens=4000,
            response_format={"type": "json_object"}
        ))

        content = response.choices[0].message.content.strip()
        result = json.loads(content)

        concept_core = result.get("concept_core_block", [])
        comparisons = result.get("comparison_tables", [])
        scenarios = result.get("clinical_scenarios", [])
        numerical = result.get("numerical_recall_points", [])
        traps = result.get("trap_patterns", [])

        if not isinstance(concept_core, list): concept_core = []
        if not isinstance(comparisons, list): comparisons = []
        if not isinstance(scenarios, list): scenarios = []
        if not isinstance(numerical, list): numerical = []
        if not isinstance(traps, list): traps = []

        return {
            "concept_core_block": concept_core[:12],
            "comparison_tables": comparisons[:4],
            "clinical_scenarios": scenarios[:5],
            "numerical_recall_points": numerical[:8],
            "trap_patterns": traps[:6],
            "generated": True,
            "model": os.getenv("AI_MODEL", "gpt-4o-mini")
        }

    except json.JSONDecodeError as e:
        print(f"JSON parse error in teaching unit: {str(e)}")
        return get_fallback_teaching_unit(subject, topic)
    except Exception as e:
        print(f"Teaching unit generation error: {str(e)}")
        return get_fallback_teaching_unit(subject, topic)


def get_fallback_teaching_unit(subject: str, topic: str) -> Dict[str, Any]:
    return {
        "concept_core_block": [
            {"title": f"Definition of {topic}", "type": "definition", "content": f"Review the fundamental definition and scope of {topic} in {subject}.", "high_yield": True},
            {"title": f"Pathophysiology of {topic}", "type": "mechanism", "content": f"Understand the underlying mechanism and pathological processes in {topic}.", "high_yield": True},
            {"title": f"Classification of {topic}", "type": "classification", "content": f"Know the standard classification systems used for {topic}.", "high_yield": True},
            {"title": f"Key Differentiations", "type": "differentiation", "content": f"Be able to differentiate between the major subtypes or similar conditions in {topic}.", "high_yield": True}
        ],
        "comparison_tables": [
            {
                "title": f"{topic} — Key Comparison",
                "columns": ["Type A", "Type B"],
                "rows": [
                    {"feature": "Etiology", "values": ["Primary", "Secondary"]},
                    {"feature": "Presentation", "values": ["Acute", "Chronic"]},
                    {"feature": "Diagnosis", "values": ["Clinical", "Lab-based"]},
                    {"feature": "Treatment", "values": ["Medical", "Surgical"]}
                ]
            }
        ],
        "clinical_scenarios": [
            {
                "scenario": f"A patient presents with classic features of {topic}. What is the most likely diagnosis?",
                "key_concept": f"Recognition of {topic} presentation",
                "expected_answer": f"Classic {topic} diagnosis",
                "teaching_point": f"Know the pathognomonic features of {topic}."
            }
        ],
        "numerical_recall_points": [
            {"fact": f"Key lab value for {topic}: review normal ranges", "context": "Essential for diagnosis", "mnemonic": None},
            {"fact": f"Staging cutoff for {topic}: review criteria", "context": "Determines treatment approach", "mnemonic": None}
        ],
        "trap_patterns": [
            {
                "trap": f"Confusing {topic} Type A with Type B",
                "why_tempting": "Similar presentation in early stages",
                "correct_approach": "Focus on the key differentiating feature",
                "related_subtopic": topic
            }
        ],
        "generated": False,
        "model": "fallback"
    }

File: backend/routes/teaching-units.js
const express = require('express');
const router = express.Router();
const { authenticate, requireAdmin } = require('../middleware/auth');
const { db } = require('../db');
const axios = require('axios');

const AI_SERVICE_URL = process.env.AI_SERVICE_URL || 'http://localhost:8000';

router.post('/generate', authenticate, async (req, res) => {
  try {
    const { subject, topic } = req.body;

    if (!subject || !topic) {
      return res.status(400).json({ error: 'Subject and topic are required' });
    }

    const existing = await db.query(
      'SELECT * FROM teaching_unit WHERE subject = $1 AND topic = $2',
      [subject, topic]
    );

    if (existing.rows.length > 0) {
      const unit = existing.rows[0];
      return res.json({
        id: unit.id,
        subject: unit.subject,
        topic: unit.topic,
        concept_core_block: safeParse(unit.concept_core_block, []),
        comparison_tables: safeParse(unit.comparison_tables, []),
        clinical_scenarios: safeParse(unit.clinical_scenarios, []),
        numerical_recall_points: safeParse(unit.numerical_recall_points, []),
        trap_patterns: safeParse(unit.trap_patterns, []),
        generated_by: unit.generated_by,
        generated_at: unit.generated_at,
        already_existed: true
      });
    }

    const subtopicsResult = await db.query(
      `SELECT DISTINCT subtopic FROM question WHERE subject = $1 AND topic = $2 AND status = 'active' AND subtopic IS NOT NULL`,
      [subject, topic]
    );
    const subtopics = subtopicsResult.rows.map(r => r.subtopic).filter(Boolean);

    const pyqResult = await db.query(
      `SELECT stem, subtopic, yield_category FROM question
       WHERE subject = $1 AND topic = $2 AND status = 'active'
       ORDER BY CASE yield_category WHEN 'core' THEN 1 WHEN 'frequent' THEN 2 ELSE 3 END
       LIMIT 10`,
      [subject, topic]
    );

    let aiResult;
    try {
      const aiResponse = await axios.post(`${AI_SERVICE_URL}/generate-teaching-unit`, {
        subject,
        topic,
        subtopics,
        pyq_data: pyqResult.rows
      }, { timeout: 60000 });
      aiResult = aiResponse.data;
    } catch (aiErr) {
      console.error('AI service error for teaching unit:', aiErr.message);
      aiResult = getFallbackUnit(subject, topic);
    }

    const id = db.generateUUID();
    await db.query(
      `INSERT INTO teaching_unit (id, subject, topic, concept_core_block, comparison_tables, clinical_scenarios, numerical_recall_points, trap_patterns, generated_by)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
      [
        id, subject, topic,
        JSON.stringify(aiResult.concept_core_block || []),
        JSON.stringify(aiResult.comparison_tables || []),
        JSON.stringify(aiResult.clinical_scenarios || []),
        JSON.stringify(aiResult.numerical_recall_points || []),
        JSON.stringify(aiResult.trap_patterns || []),
        aiResult.model || 'unknown'
      ]
    );

    res.json({
      id,
      subject,
      topic,
      concept_core_block: aiResult.concept_core_block || [],
      comparison_tables: aiResult.comparison_tables || [],
      clinical_scenarios: aiResult.clinical_scenarios || [],
      numerical_recall_points: aiResult.numerical_recall_points || [],
      trap_patterns: aiResult.trap_patterns || [],
      generated_by: aiResult.model || 'unknown',
      generated_at: new Date().toISOString(),
      already_existed: false
    });
  } catch (error) {
    console.error('Generate teaching unit error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/regenerate', authenticate, async (req, res) => {
  try {
    const { subject, topic } = req.body;

    if (!subject || !topic) {
      return res.status(400).json({ error: 'Subject and topic are required' });
    }

    await db.query(
      'DELETE FROM teaching_unit WHERE subject = $1 AND topic = $2',
      [subject, topic]
    );

    const subtopicsResult = await db.query(
      `SELECT DISTINCT subtopic FROM question WHERE subject = $1 AND topic = $2 AND status = 'active' AND subtopic IS NOT NULL`,
      [subject, topic]
    );
    const subtopics = subtopicsResult.rows.map(r => r.subtopic).filter(Boolean);

    const pyqResult = await db.query(
      `SELECT stem, subtopic, yield_category FROM question
       WHERE subject = $1 AND topic = $2 AND status = 'active'
       ORDER BY CASE yield_category WHEN 'core' THEN 1 WHEN 'frequent' THEN 2 ELSE 3 END
       LIMIT 10`,
      [subject, topic]
    );

    let aiResult;
    try {
      const aiResponse = await axios.post(`${AI_SERVICE_URL}/generate-teaching-unit`, {
        subject,
        topic,
        subtopics,
        pyq_data: pyqResult.rows
      }, { timeout: 60000 });
      aiResult = aiResponse.data;
    } catch (aiErr) {
      console.error('AI service error for teaching unit regeneration:', aiErr.message);
      aiResult = getFallbackUnit(subject, topic);
    }

    const id = db.generateUUID();
    await db.query(
      `INSERT INTO teaching_unit (id, subject, topic, concept_core_block, comparison_tables, clinical_scenarios, numerical_recall_points, trap_patterns, generated_by)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
      [
        id, subject, topic,
        JSON.stringify(aiResult.concept_core_block || []),
        JSON.stringify(aiResult.comparison_tables || []),
        JSON.stringify(aiResult.clinical_scenarios || []),
        JSON.stringify(aiResult.numerical_recall_points || []),
        JSON.stringify(aiResult.trap_patterns || []),
        aiResult.model || 'unknown'
      ]
    );

    res.json({
      id,
      subject,
      topic,
      concept_core_block: aiResult.concept_core_block || [],
      comparison_tables: aiResult.comparison_tables || [],
      clinical_scenarios: aiResult.clinical_scenarios || [],
      numerical_recall_points: aiResult.numerical_recall_points || [],
      trap_patterns: aiResult.trap_patterns || [],
      generated_by: aiResult.model || 'unknown',
      generated_at: new Date().toISOString()
    });
  } catch (error) {
    console.error('Regenerate teaching unit error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/', authenticate, async (req, res) => {
  try {
    const { subject } = req.query;

    let query = 'SELECT * FROM teaching_unit';
    const params = [];

    if (subject) {
      query += ' WHERE subject = $1';
      params.push(subject);
    }

    query += ' ORDER BY subject, topic';

    const result = await db.query(query, params);

    const units = result.rows.map(row => ({
      id: row.id,
      subject: row.subject,
      topic: row.topic,
      concept_count: safeParse(row.concept_core_block, []).length,
      comparison_count: safeParse(row.comparison_tables, []).length,
      scenario_count: safeParse(row.clinical_scenarios, []).length,
      numerical_count: safeParse(row.numerical_recall_points, []).length,
      trap_count: safeParse(row.trap_patterns, []).length,
      generated_by: row.generated_by,
      generated_at: row.generated_at
    }));

    res.json({ units });
  } catch (error) {
    console.error('List teaching units error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/:subject/:topic', authenticate, async (req, res) => {
  try {
    const { subject, topic } = req.params;

    const result = await db.query(
      'SELECT * FROM teaching_unit WHERE subject = $1 AND topic = $2',
      [decodeURIComponent(subject), decodeURIComponent(topic)]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Teaching unit not found for this topic' });
    }

    const row = result.rows[0];
    res.json({
      id: row.id,
      subject: row.subject,
      topic: row.topic,
      concept_core_block: safeParse(row.concept_core_block, []),
      comparison_tables: safeParse(row.comparison_tables, []),
      clinical_scenarios: safeParse(row.clinical_scenarios, []),
      numerical_recall_points: safeParse(row.numerical_recall_points, []),
      trap_patterns: safeParse(row.trap_patterns, []),
      generated_by: row.generated_by,
      generated_at: row.generated_at
    });
  } catch (error) {
    console.error('Get teaching unit error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.delete('/:id', authenticate, requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;
    await db.query('DELETE FROM teaching_unit WHERE id = $1', [id]);
    res.json({ message: 'Teaching unit deleted' });
  } catch (error) {
    console.error('Delete teaching unit error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/topics-available', authenticate, async (req, res) => {
  try {
    const { subject } = req.query;

    let query = `
      SELECT q.subject, q.topic, COUNT(*) as question_count,
             SUM(CASE WHEN q.yield_category = 'core' THEN 1 ELSE 0 END) as core_count,
             tu.id as has_unit
      FROM question q
      LEFT JOIN teaching_unit tu ON q.subject = tu.subject AND q.topic = tu.topic
      WHERE q.status = 'active'`;
    const params = [];

    if (subject) {
      query += ' AND q.subject = $1';
      params.push(subject);
    }

    query += ' GROUP BY q.subject, q.topic ORDER BY q.subject, q.topic';

    const result = await db.query(query, params);

    res.json({
      topics: result.rows.map(r => ({
        subject: r.subject,
        topic: r.topic,
        question_count: parseInt(r.question_count),
        core_count: parseInt(r.core_count || 0),
        has_teaching_unit: !!r.has_unit
      }))
    });
  } catch (error) {
    console.error('Topics available error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

function safeParse(value, fallback) {
  if (!value) return fallback;
  try { return JSON.parse(value); } catch { return fallback; }
}

function getFallbackUnit(subject, topic) {
  return {
    concept_core_block: [
      { title: `Definition of ${topic}`, type: 'definition', content: `Review the fundamental definition and scope of ${topic} in ${subject}.`, high_yield: true },
      { title: `Pathophysiology`, type: 'mechanism', content: `Understand the underlying mechanism and pathological processes in ${topic}.`, high_yield: true },
      { title: `Classification`, type: 'classification', content: `Know the standard classification systems used for ${topic}.`, high_yield: true },
      { title: `Key Differentiations`, type: 'differentiation', content: `Be able to differentiate between the major subtypes or similar conditions.`, high_yield: true }
    ],
    comparison_tables: [
      {
        title: `${topic} — Key Comparison`,
        columns: ['Type A', 'Type B'],
        rows: [
          { feature: 'Etiology', values: ['Primary', 'Secondary'] },
          { feature: 'Presentation', values: ['Acute', 'Chronic'] },
          { feature: 'Diagnosis', values: ['Clinical', 'Lab-based'] },
          { feature: 'Treatment', values: ['Medical', 'Surgical'] }
        ]
      }
    ],
    clinical_scenarios: [
      {
        scenario: `A patient presents with classic features of ${topic}.`,
        key_concept: `Recognition of ${topic} presentation`,
        expected_answer: `Classic ${topic} diagnosis`,
        teaching_point: `Know the pathognomonic features of ${topic}.`
      }
    ],
    numerical_recall_points: [
      { fact: `Key lab value for ${topic}: review normal ranges`, context: 'Essential for diagnosis', mnemonic: null }
    ],
    trap_patterns: [
      {
        trap: `Confusing similar presentations in ${topic}`,
        why_tempting: 'Overlapping clinical features',
        correct_approach: 'Focus on pathognomonic differentiators',
        related_subtopic: topic
      }
    ],
    model: 'fallback'
  };
}

module.exports = router;

